### PR TITLE
Saml Cert Expiry Validation

### PR DESCRIPTION
**Pull request by Shaaali**
### Proposed changes in this pull request
$subject

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreAdmin.java [hunk 1]
```diff
@@ -62,10 +62,7 @@
 import java.security.cert.X509Certificate;
 import java.text.Format;
 import java.text.SimpleDateFormat;
-import java.util.ArrayList;
-import java.util.Enumeration;
-import java.util.Iterator;
-import java.util.List;
+import java.util.*;
```
Comment (senthalan) [inline components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreAdmin.java:65]: we can't have * imports

File: components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreAdmin.java [hunk 2]
```diff
@@ -363,6 +359,27 @@ public void importCertToStore(String fileName, String certData, String keyStoreN
...
+            throws SecurityConfigException {
+        try {
+            if (keyStoreName == null) {
+                throw new SecurityConfigException("Key Store name can't be null");
+            }
+
+            KeyStore ks = getKeyStore(keyStoreName);
+            X509Certificate cert = extractCertificate(certData);
+
+            importCertToStore(fileName,cert,keyStoreName,ks);
```
Comment (senthalan) [inline components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreAdmin.java:382]: format issue

File: components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreAdmin.java [hunk 3]
```diff
@@ -915,4 +932,17 @@ public X509Certificate extractCertificate(String certData) throws SecurityConfig
         }
         return cert;
     }
+
+    public static boolean isCertificateExpired(java.security.cert.X509Certificate certificate) {
```
Comment (senthalan) [inline components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreAdmin.java:936]:
```suggestion
    public static boolean isCertificateExpired(X509Certificate certificate) {
```

File: components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreAdmin.java [hunk 4]
```diff
@@ -340,16 +337,15 @@ public void deleteStore(String keyStoreName) throws SecurityConfigException {
...
-            throws SecurityConfigException {
+    public void importCertToStore(String fileName, X509Certificate cert, String keyStoreName, KeyStore ks) throws SecurityConfigException {
+
         try {
-            if (keyStoreName == null) {
-                throw new SecurityConfigException("Key Store name can't be null");
+            if(isCertificateExpired(cert)){
+                Exception e = new SecurityConfigException("Certificate is expired.");
+                log.error(e);
+                throw e;
```
Comment (senthalan) [inline components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreAdmin.java:346]: No need to log and throw the e

File: components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreAdmin.java [hunk 5]
```diff
@@ -363,6 +359,27 @@ public void importCertToStore(String fileName, String certData, String keyStoreN
             if (isTrustStore(keyStoreName)) {
                 System.setProperty(IdentityUtil.PROP_TRUST_STORE_UPDATE_REQUIRED, "true");
             }
+        } catch (SecurityConfigException e) {
+            throw e;
+        } catch (Exception e) {
+            String msg = "Error when importing cert to the keyStore";
+            log.error(msg, e);
+            throw new SecurityConfigException(msg, e);
```
Comment (senthalan) [inline components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreAdmin.java:367]: No need to log and throw e.
Let's add more context to the error log

File: components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreAdmin.java (hunk 4 above)
Comment (senthalan) [reply components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreAdmin.java:346]: Let's add more context to the error log

File: components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreAdmin.java [hunk 6]
```diff
@@ -340,16 +337,15 @@ public void deleteStore(String keyStoreName) throws SecurityConfigException {
         }
     }
 
-    public void importCertToStore(String fileName, String certData, String keyStoreName)
-            throws SecurityConfigException {
+    public void importCertToStore(String fileName, X509Certificate cert, String keyStoreName, KeyStore ks) throws SecurityConfigException {
```
Comment (senthalan) [inline components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreAdmin.java:340]: let's add method comment

**Comment by pamodaaw** [comment]
Thank you for your contribution and the effort you've put into this PR! 
Since it has been inactive for a while, we’re closing it for now to keep the repository organized. If you'd like to revisit this or if there's an update, feel free to reopen

**Pull request closed without merging**